[DotNetCore] Support explicit .NET Core Sdk imports

mrward opened the pull request:
Local monodevelop branch with some tweaks of the pull request #5597 to verify the tests.

Adds support for:

 - `<Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />`
- `<Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" />`
 - `<Sdk Name="Microsoft.NET.Sdk" Version="1.0.0" />`
 
Fixes VSTS #522315 - Support explicit .NET Core Sdk imports
Fixes github #5712 - Unable to load SDK-style projects that lack `Sdk` attribute

Looking at the [spec](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk) and the [MSBuild unit tests](https://github.com/Microsoft/msbuild/blob/7b3a457540ab87cf02f6d41ce6095e6ecbbb6ea7/src/Build.UnitTests/Evaluation/ProjectSdkImplicitImport_Tests.cs) it looks like the logic we currently have for Import Sdk elements needs changing. Import elements can define a Version. Also Sdk and Import elements both support a MinimumVersion attribute. Currently the Import/Sdk attribute is parsed as though it has could contain an Sdk and a Version part which looks like it is incorrect. The Project/Sdk attribute does support defining both the sdk, version and minimum version. We could fix this in a separate pull request.

mathieubourgeois committed:
Adding support for all officially supported methods of declaring an Sdk usage in .Net Core
MonoDevelop only supported the Sdk attribute on the Project node, but there is also
- Sdk node as a child of the Project node. Like the Project attribute, the imports are then implicit
- Explicit Sdk imports (<Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />). In this case, the user already placed explicit imports so we shouldn't add them in its place

mathieubourgeois committed:
Cleaned Api based on Matt Ward's recommandations
- Renamed ProjectSdk back to Sdk to represent the Sdk set in the project node; to get a string representing all Sdk in the project, no matter where they're defined, call GetProjectSdksName()
- Removed SdkWithImplicitImports, call GetReferencedSdks and check for Implicit instead
- GetReferencedSdks now returns SdkDefinition, which knows if an Sdk has been defined implicitly or explicitly
- Provide a method to actually set the Sdk for the whole project, SetProjectSdk. This is to mirror the original behavior of Sdk, which had a setter. Like the original setter, it's unused. However, this function will Clear all sdks definition from the project, to ensure that the Sdk provided is the only one set. It will be set implicitly on the project node no matter what was before though...

mathieubourgeois committed:
Adding some basic load tests for Sdk formats, checking if the Project's sdk has been correctly defined and if the number of implicit imports matches how it's been defined in the project.

mathieubourgeois committed:
Multiple extra cleanup :
- Put back GetReferencedSDKs to its already standard api
- Add a new version of GetReferencedSDKs to choose if we want implicit and/or explicit sdks
- Remove GetProjectSDKsName, replaced usage with one of the two GetReferencedSdks
- Correct DotNetCoreProjectExtension.OnBeginLoad to actually use Project Sdk instead of all referenced Sdks

mathieubourgeois committed:
Remove SetProjectSdk and ClearAllSdks; not necessary

mathieubourgeois committed:
Add missing check for sdk array generation
Could cause issue if we call GetReferencedSdks(bool, bool) before any calls to GetReferencedSdks(), returning null even if there were sdks defined

mrward committed:
[DotNetCore] Improve code formatting

Minor changes to code formatting indentation and fix mixed new lines.

mrward committed:
[DotNetCore] Add test for logic detecting a web project

The .NET Core project extension has an IsWeb property that checks
to see if Microsoft.NET.Sdk.Web is referenced.

mrward committed:
[Core] Rename new MSBuildProject method to get sdks

Rename the new GetReferencedSdks to GetImplicitlyImportedSdks and
make the method internal.

mrward committed:
[DotNetCore] Add tests to check extension used for sdk projects

Tests added that verify the .NET Core project extension is used
for SDK style projects that have explicit imports or use Sdk
MSBuild items.

Therzok commented on main/src/addins/MonoDevelop.DotNetCore/MonoDevelop.DotNetCore/DotNetCoreProjectExtension.cs:
```diff
@@ -416,7 +416,7 @@ protected override void OnBeginLoad ()
 		protected bool IsWebProject (DotNetProject project)
 		{
-			return (project.MSBuildProject.Sdk?.IndexOf ("Microsoft.NET.Sdk.Web", System.StringComparison.OrdinalIgnoreCase) ?? -1) != -1;
+			return (project.MSBuildProject.GetReferencedSDKs ()?.FirstOrDefault (x => x?.IndexOf ("Microsoft.NET.Sdk.Web", System.StringComparison.OrdinalIgnoreCase) != -1) != null);
```
can `x` be null here?

mrward replied:
Not with the current implementation. GetReferencesSdks should always at least return an empty array and should never return any null items. So I think we can get rid of the null ? operator for the entire line of code.

mrward committed:
[DotNetCore] Remove null checks which are not needed

MSBuildProject's GetReferencedSDKs will never return a null array.
It will also not include any null strings.
